### DISCUSSION TITLE
Font naming in created PDF confuses Illustrator

### DISCUSSION THREAD

**Discussion by Andreasox** [pymupdf/PyMuPDF#3088] [Looking for help] [answered]
When using this syntax:
page.insert_font(fontname='GothamNarrow-Medium',fontfile='/Users/admin/Dropbox/tkartor/gtfs/Gotham Narrow Font Family/GothamNarrow-Medium.otf')
shape.insert_text((128,32), "Medium",color=color_black,fontname='GothamNarrow-Medium', fontsize=8,oc=oc_timetable)
shape.insert_text((32,132), "Cambria",color=color_black,fontname='Cambria',fontfile='/Users/admin/Library/Fonts/Cambria.ttf', fontsize=8,oc=oc_timetable)
shape.insert_text((80,132), "Calibri-Reg",color=color_black,fontname='Calibri-Regular',fontfile='/Users/admin/Library/Fonts/Calibri Regular.ttf', fontsize=8,oc=oc_timetable)
doc.subset_fonts()

the created PDF (minimized_u.pdf, uncompressed with pdftk) validates correctly with Acrobat Preflight, and is visually correct when viewed with eg. Acrobat.
But when opening the PDF with Illustrator, we get a "missing fonts" message as Illustrator requires that fonts are installed and matches with exact naming.

minimized_u.pdf includes for example the following font names:
/BaseFont /WGOXFK+Gotham#20Narrow#20Medium#20Regular  (row 197) 
/BaseFont /XSAIXJ+Cambria#20Regular (row 209)
/BaseFont /KMZUDY+Calibri#20Regular (row 221)

These rows are manually changed to the below in minimized_in_AI.pdf, which is opened without any error by Illustrator
(i.e. Illustrator can match these names to installed fonts):
/BaseFont /WGOXFK+GothamNarrow-Medium
/BaseFont /XSAIXJ+Cambria
/BaseFont /KMZUDY+Calibri

Illustrator message when opening the file:
<img width="689" alt="image" src="https://github.com/pymupdf/PyMuPDF/assets/6737097/951443a9-f25c-4704-80f4-dd14351d84f0">
[minimized_u.pdf](https://github.com/pymupdf/PyMuPDF/files/14037798/minimized_u.pdf)
[minimized_in_AI.pdf](https://github.com/pymupdf/PyMuPDF/files/14038061/minimized_in_AI.pdf)

Our workaround is to uncompress the PDF file with pdftk and replace needed strings as described.
But some way to define font names that matches what is required would be nice. It is enough to just use the font file name.

**Comment by JorjMcKie** [comment]
Well, the font name in use is like the font's self identification: a string stored inside the font binary.
MuPDF certainly exerts no influence here.

I noticed that when you insert a font to a page, you are using a very long `fontname` parameter. I fear a misconception here:
This is the font's **_reference_** name - which is completely arbitrary and disconnected from the self-identification.
IAW you could use `fontname="hugo"`, and all would be roses. In the following item
`(5, 'n/a', 'Type1', 'Helvetica-Bold', 'hebo', 'WinAnsiEncoding')`
`hebo` represents "fontname". The file's self-identification is and remains "Helvetica-Bold" even if the fontname would have been stored as (typically) "F0".

**Reply by JorjMcKie** [reply]
The font's self-id "KMZUDY+Calibri#20Regular" actually is "KMZUDY+Calibri Regular", the "#20" represents the hex version of a space `0x20`.
Possibly, that obscure Illustrator doesn't understand that and takes the first part of the string until the first blank byte only.

**Reply by JorjMcKie** [reply]
I recommend you submit a bug report to Illustrator - if I'm right, then that's what we have here.

**Reply by Andreasox** [reply]
I agree that it certainly feels as at least a limitation in Illustrator. It seems to be reading the PS names instead of the self-identification. But it is quite confusing to understand what really goes on at the Illustrator side.
How is the font self-identification defined? Is there any tool that can be used to check it?
What I now found in FontForge as "TTF names" may be what MuPDF uses.
<img width="642" alt="image" src="https://github.com/pymupdf/PyMuPDF/assets/6737097/d475784c-bf57-4657-a5d9-ccff30602e4a">
<img width="549" alt="image" src="https://github.com/pymupdf/PyMuPDF/assets/6737097/461b8ae4-5751-49d1-a4d7-02ac589f1ad5">
Compare this with PS Names but there is still a diff for GothamNarrow-Medium versus Gotham Narrow Medium
<img width="314" alt="image" src="https://github.com/pymupdf/PyMuPDF/assets/6737097/94f0633d-8d6d-4ce0-96c1-bea47e8990d8">
<img width="410" alt="image" src="https://github.com/pymupdf/PyMuPDF/assets/6737097/33782297-b422-4773-8159-220867ef2f8a">

**Comment by JorjMcKie** [comment]
> How is the font self-identification defined? Is there any tool that can be used to check it?

Sure: Take the font file and make a `fitz.Font` from it. Like this:
```python
for item in page.get_fonts():
    xref = item[0]
    if item[1]=="n/a":  # non-embedded font
        continue
    print()
    print(item)
    ff = doc.extract_font(xref)
    buff = ff[-1]
    font = fitz.Font(fontbuffer=buff)
    print(f"{font.name=}")

    

(12, 'ttf', 'TrueType', 'FNUUTH+Calibri-Bold', 'R8', '')
font.name='Calibri Bold'

(13, 'ttf', 'TrueType', 'DOKBTG+Calibri', 'R10', '')
font.name='Calibri Regular'

(14, 'ttf', 'TrueType', 'NOHSJV+Calibri-Light', 'R12', '')
font.name='Calibri Light'

(15, 'ttf', 'TrueType', 'NZNDCL+CourierNewPSMT', 'R14', '')
font.name='Courier New Regular'

(16, 'ttf', 'Type0', 'MNCSJY+SymbolMT', 'R17', 'Identity-H')
font.name='(null)'

(17, 'cff', 'Type1', 'UAEUYH+Helvetica', 'R20', 'WinAnsiEncoding')
font.name='Nimbus Sans L Regular'

(18, 'ttf', 'Type0', 'ECPLRU+Calibri', 'R23', 'Identity-H')
font.name='(null)'

(19, 'ttf', 'Type0', 'TONAYT+CourierNewPSMT', 'R27', 'Identity-H')
font.name='(null)'
```
As you can see, not every font has a name for itself.
Adobe manual:
```python
doc=fitz.open("adobe-2008.pdf")
page=doc[0]
for item in page.get_fonts():
    if item[1] == "n/a":
        continue
    xref = item[0]
    ff = doc.extract_font(xref)
    buff = ff[-1]
    font = fitz.Font(fontbuffer=buff)
    print()
    print(item)
    print(f"{font.name=}")

    

(14, 'cff', 'Type1', 'DDPEFM+Helvetica', 'T1_0', 'WinAnsiEncoding')
font.name='Helvetica Regular'

(15, 'cff', 'Type1', 'DDPEIM+Helvetica-Bold', 'T1_1', 'WinAnsiEncoding')
font.name='Helvetica-Bold Regular'
```

**Reply by Andreasox** [reply]
Thanks. 
Using your code (still at 1.23.8 and on Macos) I get the following:
(27, 'otf', 'Type0', 'Gotham Narrow Medium', 'GothamNarrow-Medium', 'Identity-H')
font.name='Gotham Narrow Medium'
(33, 'ttf', 'Type0', 'Cambria Regular', 'Cambria', 'Identity-H')
font.name='Cambria Regular'

but in the created PDF (attached) - where do the Regular for the Gotham comes from?
/BaseFont /KSHTKE+Gotham#20Narrow#20Medium#20Regular
/BaseFont /GIZPBR+Cambria#20Regular

Gotham font:
https://www.dropbox.com/scl/fi/omdtevtwi9r6ps2ym2e44/GothamNarrow-Medium.otf?rlkey=bo7w24upjt9iopb8am51ks4ar&dl=0
[minimized_u.pdf](https://github.com/pymupdf/PyMuPDF/files/14047840/minimized_u.pdf)

**Reply by JorjMcKie** [reply]
With the last attached file and this script, I see absolutely no discrepancy:
```python
import fitz, os

print(fitz.__doc__)
files = os.listdir()
for filename in files:
    if not filename.endswith(".pdf"):
        continue
    doc = fitz.open(filename)
    page = doc[0]
    print("-" * 80)
    print(f"{filename=}")
    for item in page.get_fonts():
        print(item)
        print("BaseFont:", doc.xref_get_key(item[0], "BaseFont"))
        ff = doc.extract_font(item[0])
        buff = ff[-1]
        if buff:
            font = fitz.Font(fontbuffer=buff)
            print(f"{font.name=}")
        else:
            print("font.name=None")
        print()
```
Output
```
--------------------------------------------------------------------------------
filename='minimized_u-last.pdf'
(9, 'otf', 'Type0', 'KSHTKE+Gotham Narrow Medium Regular', 'GothamNarrow-Medium', 'Identity-H')
BaseFont: ('name', '/KSHTKE+Gotham Narrow Medium Regular')
font.name='Gotham Narrow Medium Regular'

(10, 'n/a', 'Type1', 'Helvetica', 'helv', 'WinAnsiEncoding')
BaseFont: ('name', '/Helvetica')
font.name=None

(11, 'ttf', 'Type0', 'QHKTVH+Calibri Bold', 'Calibri-Bold', 'Identity-H')
BaseFont: ('name', '/QHKTVH+Calibri Bold')
font.name='Calibri Bold'

(12, 'otf', 'Type0', 'YNVHTP+Gotham Narrow Bold Regular', 'GothamNarrow-Bold', 'Identity-H')
BaseFont: ('name', '/YNVHTP+Gotham Narrow Bold Regular')
font.name='Gotham Narrow Bold Regular'

(13, 'ttf', 'Type0', 'ISZGRN+Calibri Bold Italic', 'Calibri-BoldItalic', 'Identity-H')
BaseFont: ('name', '/ISZGRN+Calibri Bold Italic')
font.name='Calibri Bold Italic'

(14, 'ttf', 'Type0', 'GIZPBR+Cambria Regular', 'Cambria', 'Identity-H')
BaseFont: ('name', '/GIZPBR+Cambria Regular')
font.name='Cambria Regular'

(15, 'otf', 'Type0', 'SKNCXK+Gotham Narrow Book Regular', 'GothamNarrow-Book', 'Identity-H')
BaseFont: ('name', '/SKNCXK+Gotham Narrow Book Regular')
font.name='Gotham Narrow Book Regular'

(16, 'otf', 'Type0', 'OCFRUN+Gotham Narrow Light Regular', 'GothamNarrow-Light', 'Identity-H')
BaseFont: ('name', '/OCFRUN+Gotham Narrow Light Regular')
font.name='Gotham Narrow Light Regular'

(17, 'ttf', 'Type0', 'WHATWH+Calibri Regular', 'Calibri-Regular', 'Identity-H')
BaseFont: ('name', '/WHATWH+Calibri Regular')
font.name='Calibri Regular'

(18, 'n/a', 'Type1', 'Helvetica-Bold', 'hebo', 'WinAnsiEncoding')
BaseFont: ('name', '/Helvetica-Bold')
font.name=None
```

**Reply by Andreasox** [reply]
Can you attach your created PDF so we can check what is written to the PDF and if Illustrator finds the referenced fonts, if installed?
I note that our reports differ, for you Gotham is always with Regular. For me it reports always without Regular.
Macos contra Windows?

As described this morning, the reported font does not match what is written to the PDF:
Row 197 in minimized_u.pdf:
/BaseFont /KSHTKE+Gotham#20Narrow#20Medium#20Regular

(9, 'otf', 'Type0', 'Gotham Narrow Book', 'GothamNarrow-Book', 'Identity-H')
font.name='Gotham Narrow Book'

(15, 'otf', 'Type0', 'Gotham Narrow Bold', 'GothamNarrow-Bold', 'Identity-H')
font.name='Gotham Narrow Bold'

(21, 'otf', 'Type0', 'Gotham Narrow Light', 'GothamNarrow-Light', 'Identity-H')
font.name='Gotham Narrow Light'

(27, 'otf', 'Type0', 'Gotham Narrow Medium', 'GothamNarrow-Medium', 'Identity-H')
font.name='Gotham Narrow Medium'

**Reply by JorjMcKie** [reply]
I did not create any PDF, only used the ones you have attached here. `filename='minimized_u-last.pdf'` being the one you have put into this thread most recently - renamed it locally.
I am using pymupdf 1.23.16 on Windows.
```python
print(fitz.__doc__)
PyMuPDF 1.23.16: Python bindings for the MuPDF 1.23.9 library (rebased implementation).
Python 3.12 running on win32 (64-bit).
```
Also tried the script on Linux: same output. No difference either if I go back to 1.23.8 or upgrade to latest 1.23.18. - not on Windows, not on Linux.

**Reply by Andreasox** [reply]
We will test on Windows.

**Reply by Andreasox** [reply]
We do not get the same report as you neither on macos nor on Ubuntu 18. We will test on Windows but that obscure OS (...) is not really important for us. 
Attaching PDF created with PyMuPDF 1.23.18 Python 3.9 on Ubuntu18. For this PDF, the only font Illustrator can find is Helvetica, which is the only font not reported ...
PyCharm report - note that we do not get the same report as you - Helvetica, BaseFont and the internal font code (KSHTKE etc.) is missing. So we seem to have a mismatch.
(9, 'otf', 'Type0', 'Gotham Narrow Book', 'GothamNarrow-Book', 'Identity-H')
font.name='Gotham Narrow Book'

(15, 'otf', 'Type0', 'Gotham Narrow Bold', 'GothamNarrow-Bold', 'Identity-H')
font.name='Gotham Narrow Bold'

(21, 'otf', 'Type0', 'Gotham Narrow Light', 'GothamNarrow-Light', 'Identity-H')
font.name='Gotham Narrow Light'

(27, 'otf', 'Type0', 'Gotham Narrow Medium', 'GothamNarrow-Medium', 'Identity-H')
font.name='Gotham Narrow Medium'

(33, 'ttf', 'Type0', 'Cambria Regular', 'Cambria', 'Identity-H')
font.name='Cambria Regular'

(39, 'ttf', 'Type0', 'Calibri Regular', 'Calibri-Regular', 'Identity-H')
font.name='Calibri Regular'

(45, 'ttf', 'Type0', 'Calibri Bold', 'Calibri-Bold', 'Identity-H')
font.name='Calibri Bold'

(51, 'ttf', 'Type0', 'Calibri Bold Italic', 'Calibri-BoldItalic', 'Identity-H')
font.name='Calibri Bold Italic'

 If updating row 169 from:
  /BaseFont /FPTZFD+Gotham#20Narrow#20Medium#20Regular
to:
  /BaseFont /GothamNarrow-Medium
Illustraror can find that font

[minimized_ubuntu.pdf](https://github.com/pymupdf/PyMuPDF/files/14061835/minimized_ubuntu.pdf)

**Reply by JorjMcKie** [reply, marked as answer]
Well, this is what the page definition (xref 4) says in the PDF:
```
4 0 obj
<<
  /Type /Page
  /MediaBox [ 0 0 595 842 ]
  /Rotate 0
  /Resources 3 0 R
  /Parent 2 0 R
  /Contents [ 19 0 R ]
>>
endobj
```
The `/Resources` dictionary at xref 3 contains
```
3 0 obj
<<
  /Font <<
    /GothamNarrow-Book 9 0 R
    /GothamNarrow-Bold 10 0 R
    /GothamNarrow-Light 11 0 R
    /GothamNarrow-Medium 12 0 R
    /Cambria 13 0 R
    /Calibri-Regular 14 0 R
    /Calibri-Bold 15 0 R
    /Calibri-BoldItalic 16 0 R
    /helv 17 0 R
    /hebo 18 0 R
  >>
  /Properties <<
    /MC0 8 0 R
  >>
>>
endobj
```
So my script above reports all those fonts correctly and completely here - both, on Windows and Ubuntu.

```
--------------------------------------------------------------------------------
filename='minimized_ubuntu.pdf'
(9, 'otf', 'Type0', 'UHKLTT+Gotham Narrow Book Regular', 'GothamNarrow-Book', 'Identity-H')
BaseFont: ('name', '/UHKLTT+Gotham Narrow Book Regular')
font.name='Gotham Narrow Book Regular'

(10, 'otf', 'Type0', 'HRPXPN+Gotham Narrow Bold Regular', 'GothamNarrow-Bold', 'Identity-H')
BaseFont: ('name', '/HRPXPN+Gotham Narrow Bold Regular')
font.name='Gotham Narrow Bold Regular'

(11, 'otf', 'Type0', 'HVOEJS+Gotham Narrow Light Regular', 'GothamNarrow-Light', 'Identity-H')
BaseFont: ('name', '/HVOEJS+Gotham Narrow Light Regular')
font.name='Gotham Narrow Light Regular'

(12, 'otf', 'Type0', 'FPTZFD+Gotham Narrow Medium Regular', 'GothamNarrow-Medium', 'Identity-H')
BaseFont: ('name', '/FPTZFD+Gotham Narrow Medium Regular')
font.name='Gotham Narrow Medium Regular'

(13, 'ttf', 'Type0', 'TBFMLM+Cambria Regular', 'Cambria', 'Identity-H')
BaseFont: ('name', '/TBFMLM+Cambria Regular')
font.name='Cambria Regular'

(14, 'ttf', 'Type0', 'XQUNPI+Calibri Regular', 'Calibri-Regular', 'Identity-H')
BaseFont: ('name', '/XQUNPI+Calibri Regular')
font.name='Calibri Regular'

(15, 'ttf', 'Type0', 'LISKLV+Calibri Bold', 'Calibri-Bold', 'Identity-H')
BaseFont: ('name', '/LISKLV+Calibri Bold')
font.name='Calibri Bold'

(16, 'ttf', 'Type0', 'PHWMEZ+Calibri Bold Italic', 'Calibri-BoldItalic', 'Identity-H')
BaseFont: ('name', '/PHWMEZ+Calibri Bold Italic')
font.name='Calibri Bold Italic'

(17, 'n/a', 'Type1', 'Helvetica', 'helv', 'WinAnsiEncoding')
BaseFont: ('name', '/Helvetica')
font.name=None

(18, 'n/a', 'Type1', 'Helvetica-Bold', 'hebo', 'WinAnsiEncoding')
BaseFont: ('name', '/Helvetica-Bold')
font.name=None
```

Anything that ignores any of those 10 fonts is buggy.
Your list mentions xref number that are no fonts in the attached file - so obviously you are referring to a different one.
Font name prefixes like `UHKLTT+` are generated when a subset of a font has been built. This is a version containing only glyphs that actually are used. This saves file size - significantly for large original fonts.

**Reply by Andreasox** [reply]
Thanks a lot. I wondered about these initial font defs in the Resources dictionary. This explains why AffinityDesigner opens the PDF correctly. Illustrator ignores this and instead tries to find the font TBFMLM+Cambria Regular in the system ... So we should send the PDF spec to Adobe ...

**Reply by JorjMcKie** [reply]
As I have copied the xref definition directly from the file you sent me (with a text editor!), it is clear that these are things compliant with the Adobe PDF specification.

Why Adobe offers a product like Illustrator that apparently is unable to comply ... remains their secret.

**Comment by Andreasox** [comment]
A final note:
Internal (glyph optimized with doc. subset_fonts) fonts are correctly recognized by e.g. Acrobat (incl. Preflight), InDesign and Affinity Designer, according to the PDF standard.
Illustrator, however, cannot handle this, i.e. do not comply with the PDF standard.
Font name replacement is needed as described.

**Reply by JorjMcKie** [reply]
Interesting information - thank you!

**Reply by Andreasox** [reply]
and thanks a lot for your support to arrive at this quite unexpected conclusion.

**Reply by JorjMcKie** [reply]
As an unsolicited information: PyMuPDF has low-level methods to manipulate the PDF source, among other things you can set `/BaseFont` values etc.
